proposals/nullable-reference-types-specification.md

```diff
@@ -0,0 +1,266 @@
+
+For the purposes of the specification, the current `nullable_type` production is renamed to `nullable_value_type`, and a `nullable_reference_type` production is added:
+
+```antlr
+reference_type
+    : ...
+    | nullable_reference_type
+    ;
+    
+nullable_reference_type
+    : non_nullable_reference_type '?'
+    ;
+    
+non_nullable_reference_type
+    : type
+    ;
+```
+
+The `non_nullable_reference_type` in a `nullable_reference_type` must be a non-nullable reference type (class, interface, delegate or array), or a type parameter that is constrained to be a non-nullable reference type (through the `class` constraint, or a class other than `object`).
+
+Nullable reference types cannot occur in the following positions:
+
+- as a base class or interface
+- as the receiver of a `member_access`
+- as the `type` in an `object_creation_expression`
+- as the `delegate_type` in a `delegate_creation_expression`
+- as the `type` in an `is_expression`, a `catch_clause` or a `type_pattern`
+- as the `interface` in a fully qualified interface member name
+
+A warning is given on a `nullable_reference_type` where the nullable annotation context is disabled.
+
+## Nullable class constraint
+
+The `class` constraint has a nullable counterpart `class?`:
+
+```antlr
+primary_constraint
+    : ...
+    | 'class' '?'
+    ;
+```
+
+## Null-forgiving expressions
+
+The post-fix `!` operator is called the null-forgiving operator.
+
+```antlr
+primary_expression
+    : ...
+    | null_forgiving_expression
+    ;
+    
+null_forgiving_expression
+    : primary_expression '!'
+    ;
+```
+
+The `primary_expression` must be of a reference type.  
+
+## Nullable compiler directives
+
+`#nullable` directives control the nullable annotation and warning contexts.
+
+```antlr
+pp_directive
+    : ...
+    | pp_nullable
+    ;
+    
+pp_nullable
+    : whitespace? '#' whitespace? 'nullable' whitespace nullable_action pp_new_line
+    ;
+    
+nullable_action
+    : 'disable'
+    | 'enable'
+    | 'restore'
+    | 'safeonly'
+    ;
+```
+
+`#pragma warning` directives are expanded to allow changing the nullable warning context, and to allow individual warnings to be enabled on even when they're disabled by default:
+
+```antlr
+pragma_warning_body
+    : ...
+    | 'warning' whitespace nullable_action whitespace 'nullable'
+    ;
+
+warning_action
+    : ...
+    | 'enable'
+    ;
+```
+
+Note that the new form of `pragma_warning_body` uses `nullable_action`, not `warning_action`.
+
+
+# Nullable contexts
+
+Every line of source code has a *nullable annotation context* and a *nullable warning context*. These control whether nullable annotations have effect, and whether nullability warnings are given. The annotation context of a given line is either *disabled* or *enabled*. The warning context of a given line is either *disabled*, *safeonly* or *enabled*.
```

Comment:
it would be likely be good to explain how code relates to "line of source code".  Especially for things like an expression that might cross a PP boundary.  Are tehre rules that should be followed? i.e. using the starting position of an expression/statement?  the ending position?  something else entirely?

Reply:
>A `type` is deemed to occur in a given annotation context when the last token of the type is within that context.

sounds like it may be sufficient :)